Wrapper to discretize observations and actions

sparisi opened the pull request:
# Description

Wrappers to discretize continuous states and actions.
There are few environments with discrete states and actions, and it can be useful to benchmark algorithms on more of them -- especially to validate results on tabular algorithms before moving to approximate algorithsm (e.g., Q-Learning before DQN).
For example, MountainCar is a well-know hard-exploration problem, but its observation space is continuous. With the `DiscretizeObservation` wrapper, one can test tablar version of their algorithm on it as well.
The same can be done to environments with continuous actions with `DiscretizeAction`.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

pseudo-rnd-thoughts commented on gymnasium/wrappers/transform_action.py:
```diff
@@ -178,3 +179,107 @@ def __init__(
+        self.high = env.action_space.high
+        self.n_dims = self.low.shape[0]
+
+        if not (np.all(np.isfinite(self.low)) and np.all(np.isfinite(self.high))):
```
Shouldn't this be np.any(...) or np.any(...) as we can't allow any low or high to be infinite?

sparisi replied:
`not np.all(np.isfinite(...))` is equivalent to what you said. I'll switch to yours because it is clearer.

pseudo-rnd-thoughts commented on gymnasium/wrappers/transform_action.py:
```diff
@@ -178,3 +179,107 @@ def __init__(
+            assert len(bins) == self.n_dims, "bins must match action dimensions"
+            self.bins = np.array(bins)
+
+        self.bin_centers = [
```
Does this work for 2+ dim box or just 1d boxes?

sparisi replied:
Any dimension. I have added an example with LunarLander to the doc.

pseudo-rnd-thoughts reviewed (changes requested):
I like the idea but I'm unsure of the implementations. Also could you add tests for it

sparisi commented:
> I like the idea but I'm unsure of the implementations. Also could you add tests for it

Done. I added functions to revert the discretization (they return a range, since it is impossible to retrieve the exact original continuous obervation), and added tests to see that discretize + revert back stays within the bins range. 
I also checked that the spaces are discretized uniformly, and that errors are raised correctly.

I see that some already existing tests fail (from wrappers/vector) but on my PC everything passes.

Btw, I am on Windows and had to run `python -m pytest` to run the tests locally, [otherwise pytest could not import anything](https://stackoverflow.com/a/49488748/754136). This also prevented me to commit from my PC (had to use the browser). Not sure if others have ever reported this.

pseudo-rnd-thoughts commented:
> Btw, I am on Windows and had to run python -m pytest to run the tests locally, [otherwise pytest could not import anything](https://stackoverflow.com/a/49488748/754136). This also prevented me to commit from my PC (had to use the browser). Not sure if others have ever reported this.

This surprises me. Also why does windows prevent normal committing more files at the same time?

sparisi commented:
> > Btw, I am on Windows and had to run python -m pytest to run the tests locally, [otherwise pytest could not import anything](https://stackoverflow.com/a/49488748/754136). This also prevented me to commit from my PC (had to use the browser). Not sure if others have ever reported this.
> 
> This surprises me. Also why does windows prevent normal committing more files at the same time?

No idea. I use GitGui and this is the first time I am encountering this problem. I managed to fix everything in the end with commits from the browser.
Let me know if the PR is ok or if I need to do something else :)

sparisi commented on gymnasium/wrappers/transform_action.py:
```diff
@@ -178,3 +179,117 @@ def __init__(
+
+        Args:
+            env: The environment to wrap.
+            bins: int or tuple of ints (number of bins per dimension).
```
"the bins used to discretrize the space" I would read this as if passing `[0, 5, 10]` means to discretize a continuous space [0, 10]. But `bins` is the number of bins per dimension, so I would have to pass `bins=2` to split the space in half.

sparisi commented on gymnasium/wrappers/transform_action.py:
```diff
@@ -178,3 +179,117 @@ def __init__(
+            for i in range(self.n_dims)
+        ]
+
+        self.action_space = Discrete(np.prod(self.bins))
```
Yes, because of that. Is there a wrapper to flatten MultiDiscrete into Discrete? I can make the discretization wrapper more general with MultiDiscrete, and then one can further flatten it to Discrete. If such a wrapper doesn't exist, I can make it.

sparisi replied:
Sure, I think it should be easy by adding an optional flag. I'll look into it!

pseudo-rnd-thoughts reviewed (approved):
Looks good, thanks for the PR @sparisi
